File: bindings/cpp/samples/spot_callback_sample.cpp
/* SPDX-License-Identifier: MPL-2.0 */

#include "sample_common.hpp"

namespace {

struct callback_result_t
{
    std::string topic;
    std::string payload;
};

void subscribe_callback (const zlink_routing_id_t *,
                         const char *topic_,
                         size_t topic_len_,
                         zlink_msg_t *parts_,
                         size_t part_count_,
                         void *userdata_)
{
    std::promise<callback_result_t> *result_promise =
      static_cast<std::promise<callback_result_t> *> (userdata_);
    assert (result_promise != NULL);
    assert (topic_ != NULL);
    assert (part_count_ == 1);

    callback_result_t result;
    result.topic.assign (topic_, topic_len_);
    result.payload.assign (
      static_cast<const char *> (zlink_msg_data (&parts_[0])),
      zlink_msg_size (&parts_[0]));

    zlink_multipart_close (parts_, part_count_);
    result_promise->set_value (result);
}

} // namespace

int main ()
{
    zlink::context_t ctx;
    zlink::service::spot_node_t pub_node (ctx);
    zlink::service::spot_node_t sub_node (ctx);
    assert (pub_node.valid ());
    assert (sub_node.valid ());

    zlink::service::spot_t pub_spot (pub_node);
    zlink::service::spot_t sub_spot (sub_node);
    assert (pub_spot.valid ());
    assert (sub_spot.valid ());

    assert (pub_node.bind ("tcp://127.0.0.1:0") == 0);
    const std::string endpoint = pub_node.last_endpoint ();
    assert (!endpoint.empty ());
    assert (sub_node.connect_peer (endpoint) == 0);

    const std::string topic = detail::k_spot_topic;
    std::promise<callback_result_t> result_promise;
    std::future<callback_result_t> result_future = result_promise.get_future ();
    assert (sub_spot.on_subscribe (&subscribe_callback, &result_promise)
            == 0);
    assert (sub_spot.set_subscription (topic) == 0);

    const std::string sent = detail::k_spot_payload;
    zlink::message_t outbound = detail::make_message (sent);
    pub_spot.publish (topic, outbound);

    const callback_result_t result = detail::wait_future (result_future, 10000);
    assert (result.topic == topic);
    assert (result.payload == detail::k_spot_payload);
    std::printf (
      "[spot/callback] publish: \"%s/%s\" → subscribe: \"%s/%s\"\n",
      topic.c_str (), sent.c_str (), result.topic.c_str (), result.payload.c_str ());
    assert (sub_spot.close () == 0);
    assert (pub_spot.close () == 0);
    assert (sub_node.close () == 0);
    assert (pub_node.close () == 0);
    return 0;
}

File: bindings/node/samples/spot_callback_sample.ts
// SPDX-License-Identifier: MPL-2.0

'use strict';

const assert = require('node:assert/strict');
const { once } = require('node:events');
const net = require('node:net');
const zlink = require('../dist');

const topic = 'room:lobby';
const sent = 'hello-spot';

async function reservePort() {
  const server = net.createServer();
  server.listen(0, '127.0.0.1');
  await once(server, 'listening');
  const { port } = server.address();
  await new Promise((resolve, reject) => server.close((error) => error ? reject(error) : resolve()));
  return port;
}

async function main() {
  const endpoint = `tcp://127.0.0.1:${await reservePort()}`;
  const ctx = new zlink.Context();
  const pubNode = new zlink.SpotNode(ctx);
  const subNode = new zlink.SpotNode(ctx);
  const pub = new zlink.Spot(pubNode);
  const sub = new zlink.Spot(subNode);

  try {
    const receivedPromise = new Promise((resolve, reject) => {
      try {
        sub.onSubscribe((routingId, receivedTopic, parts) => {
          resolve({ routingId, receivedTopic, parts });
        });
      } catch (error) {
        reject(error);
      }
    });
    const timeoutPromise = new Promise((_, reject) => {
      setTimeout(() => reject(new Error('spot callback sample timed out')), 5000);
    });

    pubNode.bind(endpoint);
    subNode.connectPeer(endpoint);
    sub.setSubscription(topic);
    const publishTimer = setInterval(() => {
      pub.publish(topic, Buffer.from(sent));
    }, 25);
    const received = await Promise.race([receivedPromise, timeoutPromise]).finally(() => {
      clearInterval(publishTimer);
    });
    assert.ok(received.routingId === null || Buffer.isBuffer(received.routingId));
    assert.equal(received.receivedTopic, topic);
    const recv = received.parts[0].data.toString();
    assert.equal(recv, sent);
    console.log(`[spot/callback] publish: "${topic}/${sent}" \u2192 subscribe: "${topic}/${recv}"`);
  } finally {
    sub.close();
    pub.close();
    subNode.close();
    pubNode.close();
    ctx.close();
  }

  process.exit(0);
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
